xds/internal/xdsclient/tests/ads_stream_watch_test.go

```diff
@@ -0,0 +1,197 @@
+func (s) TestADS_WatchState_StreamBreaks(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Create an xDS management server with a restartable listener.
+	l, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("Failed to create a local listener for the xDS management server: %v", err)
+	}
+	lis := testutils.NewRestartableListener(l)
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{Listener: lis})
+
+	// Create an xDS client with bootstrap pointing to the above server.
+	nodeID := uuid.New().String()
+	bc := e2e.DefaultBootstrapContents(t, nodeID, mgmtServer.Address)
+	testutils.CreateBootstrapFileForTesting(t, bc)
+	client := createXDSClient(t, bc)
+
+	// Create a watch for the first listener resource and verify that the timer
+	// is running and the watch state is `requested`.
+	const listenerName1 = "listener1"
+	ldsCancel1 := xdsresource.WatchListener(client, listenerName1, noopListenerWatcher{})
+	defer ldsCancel1()
+	if err := waitForResourceWatchState(ctx, client, listenerName1, ads.ResourceWatchStateRequested, true); err != nil {
+		t.Fatal(err)
+	}
+
+	// Configure the first resource on the management server. This should result
+	// in the resource being pushed to the xDS client and should result in the
+	// timer getting stopped and the watch state moving to `received`.
+	const routeConfigName = "route-config"
+	listenerResource1 := e2e.DefaultClientListener(listenerName1, routeConfigName)
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{listenerResource1},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+	if err := waitForResourceWatchState(ctx, client, listenerName1, ads.ResourceWatchStateReceived, false); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a watch for the second listener resource and verify that the timer
+	// is running and the watch state is `requested`.
+	const listenerName2 = "listener2"
+	ldsCancel2 := xdsresource.WatchListener(client, listenerName2, noopListenerWatcher{})
+	defer ldsCancel2()
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateRequested, true); err != nil {
+		t.Fatal(err)
+	}
+
+	// Stop the server to break the ADS stream. Since the first resource was
+	// already received, this should not change anything for it. But for the
+	// second resource, it should result in the timer getting stopped and the
+	// watch state moving to `started`.
+	lis.Stop()
+	if err := waitForResourceWatchState(ctx, client, listenerName2, ads.ResourceWatchStateStarted, false); err != nil {
+		t.Fatal(err)
+	}
+	if err := verifyResourceWatchState(client, listenerName1, ads.ResourceWatchStateReceived, false); err != nil {
+		t.Fatal(err)
+	}
+
+	// Restart the server and verify that the timer is running and the watch
+	// state is `requested`.
+	lis.Restart()
```

Comment:
Optional: check the resource watch state of the first one is the same? And then update that one too and see if it picks up change.

Reply:
Added a check here to ensure that the watch state does not change after the ADS stream is restored. I didn't add a check for the update because there is nothing interesting that happens on a state of a resource after it has been received *once*.

Reply:
Alright sounds good.